include/__generator.hpp

```diff
@@ -368,18 +378,33 @@ struct __generator_promise_base
     using __reference_decayed = std::remove_cvref_t<__reference>;
 
     std::suspend_always yield_value(__reference_decayed & __x) noexcept {
-        __root_->__value_ = std::addressof(__x);
+        if constexpr (__is_cheap) {
+            __root_->__value_ = __x;
+        }
+        else {
+            __root_->__value_ = std::addressof(__x);
+        }
         return {};
     }
 
     std::suspend_always yield_value(__reference_decayed&& __x) noexcept  {
-        __root_->__value_ = std::addressof(__x);
+        if constexpr (__is_cheap) {
+            __root_->__value_ = __x;
+        }
+        else {
+            __root_->__value_ = std::addressof(__x);
+        }
         return {};
     }
 
     std::suspend_always yield_value(const __reference_decayed & __x) noexcept
     requires (std::is_const_v<std::remove_reference_t<__reference>>){
-        __root_->__value_ = std::addressof(__x);
+        if constexpr (__is_cheap) {
+            __root_->__value_ = __x;
+        }
+        else {
+            __root_->__value_ = std::addressof(__x);
+        }
         return {};
     }
```

Comment:
That is the whole point, after talking to Mathias ( note we still don't agree on everything but I think he is right on that point) - it is less work and more correct to let types bind to the const T& when a copy needs to be made.
It acts more like a regular function call, still do a copy excepts the compiler deals with it ( so I guess it might be a bit less pointer chasing, although you'd know better than me.

It also weeds out types with an explicit constructor, which I think is a feature.